File: src/ChatApp.Infrastructure/Storage/LocalFileStorage.cs
using System.Security.Cryptography;
using ChatApp.Application.Interfaces;
using ChatApp.Domain.Enums;
using ChatApp.Infrastructure.Persistence;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace ChatApp.Infrastructure.Storage;

public class LocalFileStorage : IFileStorage
{
    private readonly string _root;
    private readonly ILogger<LocalFileStorage> _log;
    private const string ChunksSubDir = "_chunks";

    public LocalFileStorage(IConfiguration cfg, ILogger<LocalFileStorage> log)
    {
        var configured = cfg["FileStorage:Root"];
        _root = string.IsNullOrWhiteSpace(configured)
            ? Path.Combine(AppContext.BaseDirectory, "uploads")
            : configured!;

        if (!Path.IsPathRooted(_root)) _root = Path.GetFullPath(_root);
        Directory.CreateDirectory(_root);
        _log = log;
    }

    public string GetStorageRoot() => _root;

    private string Resolve(string relativePath)
    {
        var full = Path.GetFullPath(Path.Combine(_root, relativePath));
        if (!full.StartsWith(_root, StringComparison.OrdinalIgnoreCase))
            throw new UnauthorizedAccessException("Path traversal detected.");
        return full;
    }

    public async Task<string> SaveAsync(Stream stream, string fileName, string contentType, string subDirectory, CancellationToken ct = default)
    {
        var safeSub = SanitizeSubDir(subDirectory);
        var dir = Path.Combine(_root, safeSub);
        Directory.CreateDirectory(dir);

        var ext = Path.GetExtension(fileName);
        var stored = $"{Guid.NewGuid():N}{ext}";
        var full = Path.Combine(dir, stored);

        await using (var fs = new FileStream(full, FileMode.CreateNew, FileAccess.Write, FileShare.None, 81920, useAsync: true))
        {
            await stream.CopyToAsync(fs, 81920, ct);
        }
        return Path.Combine(safeSub, stored).Replace('\\', '/');
    }

    public Task<Stream> OpenReadAsync(string relativePath, CancellationToken ct = default)
    {
        var full = Resolve(relativePath);
        if (!File.Exists(full)) throw new FileNotFoundException("File not found.", full);
        return Task.FromResult<Stream>(new FileStream(full, FileMode.Open, FileAccess.Read, FileShare.Read, 81920, useAsync: true));
    }

    public Task DeleteAsync(string relativePath, CancellationToken ct = default)
    {
        var full = Resolve(relativePath);
        if (File.Exists(full)) File.Delete(full);
        return Task.CompletedTask;
    }

    public Task<bool> ExistsAsync(string relativePath, CancellationToken ct = default)
    {
        return Task.FromResult(File.Exists(Resolve(relativePath)));
    }

    public async Task<string> SaveChunkAsync(Stream chunkStream, string chunkDirectory, int chunkIndex, CancellationToken ct = default)
    {
        var dir = Resolve(chunkDirectory);
        Directory.CreateDirectory(dir);
        var chunkPath = Path.Combine(dir, $"{chunkIndex:D8}.part");

        await using (var fs = new FileStream(chunkPath, FileMode.Create, FileAccess.Write, FileShare.None, 81920, useAsync: true))
        {
            await chunkStream.CopyToAsync(fs, 81920, ct);
        }
        return chunkPath;
    }

    public async Task MergeChunksAsync(string chunkDirectory, string outputPath, int totalChunks, CancellationToken ct = default)
    {
        var srcDir = Resolve(chunkDirectory);
        var outPath = Resolve(outputPath);
        Directory.CreateDirectory(Path.GetDirectoryName(outPath)!);

        await using (var dest = new FileStream(outPath, FileMode.Create, FileAccess.Write, FileShare.None, 81920, useAsync: true))
        {
            for (var i = 0; i < totalChunks; i++)
            {
                var chunk = Path.Combine(srcDir, $"{i:D8}.part");
                if (!File.Exists(chunk))
                    throw new InvalidOperationException($"Chunk {i} missing in upload.");

                using var src = new FileStream(chunk, FileMode.Open, FileAccess.Read, FileShare.Read, 81920, useAsync: true);
                await src.CopyToAsync(dest, 81920, ct);
            }
        }
    }

    public Task CleanupChunksAsync(string chunkDirectory, CancellationToken ct = default)
    {
        var dir = Resolve(chunkDirectory);
        if (Directory.Exists(dir))
        {
            try { Directory.Delete(dir, recursive: true); } catch { /* ignore */ }
        }
        return Task.CompletedTask;
    }

    public Task DeleteChunkAsync(string chunkDirectory, int chunkIndex, CancellationToken ct = default)
    {
        var dir = Resolve(chunkDirectory);
        var chunk = Path.Combine(dir, $"{chunkIndex:D8}.part");
        if (File.Exists(chunk)) File.Delete(chunk);
        return Task.CompletedTask;
    }

    public async Task<long> GetFileSizeAsync(string relativePath, CancellationToken ct = default)
    {
        var full = Resolve(relativePath);
        var fi = new FileInfo(full);
        return fi.Exists ? fi.Length : 0;
    }

    public async Task<(int ReceivedChunks, bool[] ChunkMap)> GetChunkMapAsync(string chunkDirectory, int totalChunks, CancellationToken ct = default)
    {
        var dir = Resolve(chunkDirectory);
        var map = new bool[totalChunks];
        var received = 0;
        if (Directory.Exists(dir))
        {
            for (var i = 0; i < totalChunks; i++)
            {
                if (File.Exists(Path.Combine(dir, $"{i:D8}.part")))
                {
                    map[i] = true;
                    received++;
                }
            }
        }
        return (received, map);
    }

    public string GetPublicUrl(string relativePath) => $"/api/files/raw?path={Uri.EscapeDataString(relativePath)}";

    public string GetThumbnailUrl(string? thumbnailPath) =>
        string.IsNullOrEmpty(thumbnailPath) ? string.Empty : GetPublicUrl(thumbnailPath);

    private static string SanitizeSubDir(string sub)
    {
        var parts = sub.Split('/', '\\', StringSplitOptions.RemoveEmptyEntries);
        var cleaned = parts.Select(p => new string(p.Where(ch => char.IsLetterOrDigit(ch) || ch == '_' || ch == '-').ToArray()));
        return string.Join('/', cleaned);
    }

    public AttachmentType DetectAttachmentType(string fileName, string contentType)
    {
        var ext = Path.GetExtension(fileName).ToLowerInvariant();
        var ctLower = (contentType ?? string.Empty).ToLowerInvariant();
        return ext switch
        {
            ".jpg" or ".jpeg" or ".png" or ".gif" or ".webp" or ".bmp" => AttachmentType.Image,
            ".mp4" or ".mov" or ".avi" or ".mkv" or ".webm" => AttachmentType.Video,
            ".mp3" or ".wav" or ".ogg" or ".m4a" or ".flac" => AttachmentType.Audio,
            ".pdf" => AttachmentType.Pdf,
            ".zip" => AttachmentType.Zip,
            ".rar" => AttachmentType.Rar,
            ".doc" or ".docx" => AttachmentType.Document,
            ".xls" or ".xlsx" => AttachmentType.Spreadsheet,
            ".ppt" or ".pptx" => AttachmentType.Presentation,
            ".txt" or ".md" => AttachmentType.Text,
            _ when ctLower.StartsWith("image/") => AttachmentType.Image,
            _ when ctLower.StartsWith("video/") => AttachmentType.Video,
            _ when ctLower.StartsWith("audio/") => AttachmentType.Audio,
            _ => AttachmentType.Other
        };
    }
}

File: backend/src/files/storage/local-file-storage.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as fs from 'fs/promises';
import * as path from 'path';
import { createReadStream, createWriteStream } from 'fs';
import { pipeline } from 'stream/promises';
import { IFileStorage } from './file-storage.interface';

const ALLOWED_EXTENSIONS = new Set([
  '.jpg', '.jpeg', '.png', '.gif', '.webp', '.bmp',
  '.mp4', '.mov', '.avi', '.mkv', '.webm',
  '.mp3', '.wav', '.ogg', '.m4a', '.flac',
  '.pdf', '.zip', '.rar',
  '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx',
  '.txt', '.md',
]);

@Injectable()
export class LocalFileStorage implements IFileStorage {
  private readonly logger = new Logger('LocalFileStorage');
  private readonly root: string;

  constructor(config: ConfigService) {
    const dir = config.get<string>('UPLOAD_DIR', './uploads');
    this.root = path.isAbsolute(dir) ? dir : path.resolve(process.cwd(), dir);
    fs.mkdir(this.root, { recursive: true }).catch((err) => {
      this.logger.error(`Failed to create upload dir ${this.root}: ${err.message}`);
    });
  }

  async saveChunk(stream: NodeJS.ReadableStream, chunkDirectory: string, chunkIndex: number): Promise<void> {
    const dir = this.resolve(chunkDirectory);
    await fs.mkdir(dir, { recursive: true });
    const chunkPath = path.join(dir, `${String(chunkIndex).padStart(8, '0')}.part`);
    const ws = createWriteStream(chunkPath, { flags: 'w' });
    await pipeline(stream, ws);
  }

  async mergeChunks(chunkDirectory: string, targetStorageKey: string, totalChunks: number): Promise<void> {
    const srcDir = this.resolve(chunkDirectory);
    const targetPath = this.resolve(targetStorageKey);
    await fs.mkdir(path.dirname(targetPath), { recursive: true });

    const ws = createWriteStream(targetPath, { flags: 'w' });
    try {
      for (let i = 0; i < totalChunks; i++) {
        const chunkPath = path.join(srcDir, `${String(i).padStart(8, '0')}.part`);
        try {
          const stat = await fs.stat(chunkPath);
          if (!stat.isFile()) throw new Error(`Chunk ${i} missing`);
        } catch {
          throw new Error(`Chunk ${i} missing in ${chunkDirectory}`);
        }
        const rs = createReadStream(chunkPath);
        await pipeline(rs, ws, { end: false });
      }
    } finally {
      ws.end();
      await new Promise<void>((resolve, reject) => {
        ws.on('finish', () => resolve());
        ws.on('error', reject);
      });
    }
  }

  openRead(storageKey: string): NodeJS.ReadableStream {
    const full = this.resolve(storageKey);
    return createReadStream(full);
  }

  async getSize(storageKey: string): Promise<number> {
    try {
      const stat = await fs.stat(this.resolve(storageKey));
      return stat.size;
    } catch { return 0; }
  }

  async delete(storageKey: string): Promise<void> {
    try { await fs.unlink(this.resolve(storageKey)); } catch { /* ignore */ }
  }

  async cleanupChunks(chunkDirectory: string): Promise<void> {
    try {
      await fs.rm(this.resolve(chunkDirectory), { recursive: true, force: true });
    } catch (err) {
      this.logger.warn(`cleanupChunks: ${err}`);
    }
  }

  async deleteChunk(chunkDirectory: string, chunkIndex: number): Promise<void> {
    try {
      await fs.unlink(path.join(this.resolve(chunkDirectory), `${String(chunkIndex).padStart(8, '0')}.part`));
    } catch { /* ignore */ }
  }

  async getChunkMap(chunkDirectory: string, totalChunks: number): Promise<{ received: number; map: boolean[] }> {
    const map = new Array(totalChunks).fill(false);
    let received = 0;
    try {
      const files = await fs.readdir(this.resolve(chunkDirectory));
      for (const f of files) {
        const m = f.match(/^(\d{8})\.part$/);
        if (m) {
          const idx = parseInt(m[1], 10);
          if (idx >= 0 && idx < totalChunks) {
            map[idx] = true;
            received++;
          }
        }
      }
    } catch { /* directory doesn't exist yet */ }
    return { received, map };
  }

  async exists(storageKey: string): Promise<boolean> {
    try {
      const stat = await fs.stat(this.resolve(storageKey));
      return stat.isFile();
    } catch { return false; }
  }

  generateStorageKey(originalFileName: string, _mimeType: string): string {
    const ext = path.extname(originalFileName).toLowerCase();
    const safeExt = ALLOWED_EXTENSIONS.has(ext) ? ext : '';
    const id = require('crypto').randomUUID().replace(/-/g, '');
    const yyyy = new Date().getFullYear();
    const mm = String(new Date().getMonth() + 1).padStart(2, '0');
    return path.posix.join('attachments', `${yyyy}`, `${mm}`, `${id}${safeExt}`);
  }

  detectType(fileName: string, mimeType: string): string {
    const ext = path.extname(fileName).toLowerCase();
    const map: Record<string, string> = {
      '.jpg': 'IMAGE', '.jpeg': 'IMAGE', '.png': 'IMAGE', '.gif': 'IMAGE', '.webp': 'IMAGE', '.bmp': 'IMAGE',
      '.mp4': 'VIDEO', '.mov': 'VIDEO', '.avi': 'VIDEO', '.mkv': 'VIDEO', '.webm': 'VIDEO',
      '.mp3': 'AUDIO', '.wav': 'AUDIO', '.ogg': 'AUDIO', '.m4a': 'AUDIO', '.flac': 'AUDIO',
      '.pdf': 'PDF',
      '.zip': 'ZIP',
      '.rar': 'RAR',
      '.doc': 'DOCUMENT', '.docx': 'DOCUMENT',
      '.xls': 'SPREADSHEET', '.xlsx': 'SPREADSHEET',
      '.ppt': 'PRESENTATION', '.pptx': 'PRESENTATION',
      '.txt': 'TEXT', '.md': 'TEXT',
    };
    if (map[ext]) return map[ext];
    const mt = (mimeType || '').toLowerCase();
    if (mt.startsWith('image/')) return 'IMAGE';
    if (mt.startsWith('video/')) return 'VIDEO';
    if (mt.startsWith('audio/')) return 'AUDIO';
    return 'OTHER';
  }

  /**
   * Resolve a storage key to an absolute filesystem path, guarding against path traversal.
   */
  private resolve(storageKey: string): string {
    const full = path.resolve(this.root, storageKey);
    if (!full.startsWith(this.root + path.sep) && full !== this.root) {
      throw new Error('Path traversal detected.');
    }
    return full;
  }
}
